cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingPollingService.java

```diff
@@ -111,7 +113,7 @@ protected final TopicId getTopicId() {
    * @throws Exception if there is error processing messages.
    */
   @Nullable
-  protected abstract String processMessages(Iterator<ImmutablePair<String, T>> messages) throws Exception;
+  protected abstract String processMessages(PeekingIterator<ImmutablePair<String, T>> messages) throws Exception;
```

Comment:
instead of forcing it to be a PeekingIterator here, I think it's cleaner if the MessageTrackingIterator wraps its input iterator into a peeking one.